include/seastar/http/client.hh

```diff
@@ -154,6 +154,23 @@ public:
     virtual ~connection_factory() {}
 };
 
+/**
+ * \brief Parameters for retrying failed requests
+ *
+ * In case request fails due to transport errors, the client may retry it on the caller's
+ * bahalf and this param structure controls the way these attempts are made
+ *
+ * \param max_count -- the maximum number of retry attempts. Default is zero, so no
+ *                     retries would happen
+ * \param pause_base -- client uses exponentially increasing delays between retry attempts
+ *                      and this is the base value for the exponent. First attempt would
+ *                      happen after this duration. Default is 100ms
```

Comment:
Since the exponent is not configurable, please at least say what it is (I looked at the code and saw it's 2).
To be honest, I'm not sure why exponential backoff makes sense for the case of a **transport error**. Presumably, this happens in the middle of a long request. Does it make sense to wait 0.1 seconds and if it happens again 10 seconds later, wait 0.2 seconds, and if it happens again later, wait 0.4 seconds? I'm not sure how this is helpful.

Please also say what a "transport error" is. If the remote host is down (either timeout in SYN, or ICMP host-not-reachable, or I don't know what) - will the request be retried or not?

Reply:
Updated comment

Well, this is what happens with S3 server (as I see it from logs, unfortunately I didn't collect strong statistics)

1. client accumulates 5 connections in pool
2. user sends a request, client picks a connection from pool and starts writing request into it
3. if writing request _or_ reading response fails with, respectively, EPIPE or ECONRESET -> goto 5
4. return success
5. client waits `0.1 << attempt` sec, then picks up another connection from pool -> goto step 2


Some heuristics:

- S3 server is pretty unstable in terms of how actively it kills the connections runtime. I could run a test for minutes using hundreds of connections and uploading gigs of data, no EPIPE/ECONRESET. But at times server could close every 2nd connection, so retry code was pretty active
- In most of the cases _single_ retry (i.e. -- using two connections) after 100ms helped
- Rarely server fails the first retry attempt (after failing the initial attempt), so client uses 3 connections to succeed
- Using first retry timeout of 10ms increases the chance of client draining the whole pool and proceeding with fresh new connection. Note, that I used 4 retries count, so 10ms + 20ms + 40ms + 80ms = 150ms.

The last observation makes me think that with S3 100ms is some sane duration after which server heals itself

Reply:
Another retry option that also worked, was -- once connection from pool closes in the middle, just go ahead and establish new one even if there are other connections in the pool

pros:
- the logic is more straightforward and doesn't imply magic constatnts

cons:
- initiates connect()ion storm at times
- the resulting patch is little bit more hairy

Reply:
@nyh , the force-new-connection options from above is in the https://github.com/xemul/seastar/tree/br-http-restart-requests-a branch. If you think it's more appropriate than the exponential-backoff-retry approach, let me know, I'll force push that branch here for review